[MRG+1] Added support for sparse multilabel y for Nearest neighbor classifiers

massich opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#Contributing-Pull-Requests
-->
#### Reference Issue
<!-- Example: Fixes #1234 -->

Fixes #8057


#### What does this implement/fix? Explain your changes.

Added support for sparse multilabel y for nearest neighbor classifiers. Firstly, it checks if the input to fit is sparse + multilabel and converts it to a dense one for storing. Also, it stores another parameter indicating whether the original input was sparse + multilabel or not. Now, in predict, if this stored value is true, then it converts y_pred to sparse CSC.
Also, added tests for the same.

#### Any other comments?

This PR wraps up @dalmia 's work in #8096

dalmia committed:
ENH: added support for sparse multilabel y

dalmia committed:
FIX: fix typo

dalmia committed:
FIX: fixed indentation error

dalmia committed:
ENH: avoid dense conversion for support

dalmia committed:
ENH: adding separate codepath for sparse multilabel case

dalmia committed:
ENH: added sparsity check to existing variable

dalmia committed:
ENH: added support for RadiusNeighborClassifier

dalmia committed:
FIX: removed unused import

dalmia committed:
FIX: replaced csr_matrix with csc_matrix in tests

dalmia committed:
TST: use assert_array_equal

dalmia committed:
FIX: replaced neigh_ind with inliers

dalmia committed:
FIX: using only inliers for calculating labels

dalmia committed:
FIX: added failure for scipy < 0.13

dalmia committed:
FIX: remove flake8 errors

dalmia committed:
FIX: moved version check to fit

dalmia committed:
TST: removed unrequired brackets

dalmia committed:
FIX: added function to calculate mode

dalmia committed:
FIX: fixes based on review

dalmia committed:
FIX: removed unused imports

Commit:
Addressed SOME of joel's comments

massich commented on sklearn/neighbors/base.py:
```diff
@@ -772,15 +772,24 @@ def fit(self, X, y):
+        if issparse(y) and self.outputs_2d_:
+            self.outputs_2d_ = 'sparse'
+            self._y = y
+            self.classes_ = [np.array([0, 1], dtype=np.int)] * y.shape[1]
```
@jnothman where )you asking for something like `self.classes_ = set(y)` ?

jnothman replied:
I just mean that, even if (for no particularly good reason) we only support y being sparse when it is multilabel (not general multioutput) someone might get a surprise by this assumption that it's all binary.

You can't just do set(y). if y is CSC or CSR, checking that `np.all(np.union1d(y.data, [0, 1]) == [0, 1])` and raising an error if it fails might help someone unsuspecting.

jnothman replied:
but it may not be necessary. if you fix it, it needs a test.

massich commented on sklearn/neighbors/classification.py:
```diff
@@ -364,27 +381,49 @@ def predict(self, X):
+                             in zip(pred_labels, weights[inliers])],
+                            dtype=np.int)
+
+        mode = mode.ravel()
```
@jnothman is `mode = mode.ravel()` right ?

glemaitre replied:
I would be careful here. We should document that input should always be 2D (to use `axis=0` from the `mode` function` and ravel will allow to always return 1D array.

jnothman commented on sklearn/neighbors/classification.py:
```diff
@@ -161,7 +161,7 @@ def predict(self, X):
             for k, classes_k in enumerate(classes_):
                 mode = self._mode(_y[neigh_ind, k].toarray(), weights)
-                y_pred_sparse_multilabel.append(csr_matrix(mode).T)
+                y_pred_sparse_multilabel.append(csc_matrix(mode).T)
```
This isn't right. You don't want to transpose the CSC. You want to build a CSC on the reshaped mode. Transposing on a numpy array is very cheap. Transposing a scipy.sparse matrix is more expensive (except for COO format).

jnothman replied:
I'd hope that tests don't pass...

massich replied:
Cool `y_pred_sparse_multilabel.append(csc_matrix(mode.T))`

Which test in `sklearn/cluster/tests/` should break when using `csc_matrix(mode).T`? I run `nosetests sklearn/cluster/` and all tests pass. Shall we add a regression test somwhere?

jnothman commented on sklearn/neighbors/classification.py:
```diff
@@ -364,27 +381,49 @@ def predict(self, X):
+                             in zip(pred_labels, weights[inliers])],
+                            dtype=np.int)
+
+        mode = mode.ravel()
```
yes, it's fine.

jnothman replied:
Now it's not clear to me why it's necessary, i.e. why mode is not already 1d

massich replied:
`ravel()` changes the shape of `mode` from `(_,1)` to `(_,)`. Removing it brakes lots of tests. Shall we remove it and fix the tests?

Commit:
Address jnothman comments

massich commented on sklearn/neighbors/tests/test_neighbors.py:
```diff
@@ -503,6 +504,53 @@ def test_kneighbors_classifier_sparse(n_samples=40,
+    for name in CLASSIFIERS:
+        yield check_classifier_sparse_multilabel_y, name
+
+def test_sparse_multilabel_y():
```
@jnothman is this the test you meant?

jnothman replied:
Remind me where I mentioned a test

massich replied:
@jnothman [here](https://github.com/scikit-learn/scikit-learn/pull/9059/#discussion_r120882985)

jnothman commented:
i don't think mode.T will actually work because mode is 1d

On 9 Jun 2017 12:54 am, "Joan Massich" <notifications@github.com> wrote:

> *@massich* commented on this pull request.
> ------------------------------
>
> In sklearn/neighbors/tests/test_neighbors.py
> <https://github.com/scikit-learn/scikit-learn/pull/9059#discussion_r120910583>
> :
>
> > +    clf.fit(X_train, y_train)
> +    y_pred = clf.predict(X_test)
> +
> +    y_sparse = csc_matrix(y)
> +    X_train, X_test, y_train, y_test = train_test_split(X, y_sparse,
> +                                                        random_state=0)
> +    clf.fit(X_train, y_train)
> +    y_sparse_pred = clf.predict(X_test)
> +    assert_array_equal(y_pred, y_sparse_pred.toarray())
> +
> +
> +def test_classifiers_sparse_multilabel_y():
> +    for name in CLASSIFIERS:
> +        yield check_classifier_sparse_multilabel_y, name
> +
> +def test_sparse_multilabel_y():
>
> @jnothman <https://github.com/jnothman> is this the test you meant?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/9059#pullrequestreview-42915463>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz69sr9kBizi_tTDz8g-GZpNQTfG6kks5sCAsmgaJpZM4N0Bnb>
> .
>

massich committed:
Merge branch 'master' into 8057

jnothman commented on sklearn/neighbors/tests/test_neighbors.py:
```diff
@@ -503,6 +504,53 @@ def test_kneighbors_classifier_sparse(n_samples=40,
+
+    clf = neighbors.KNeighborsClassifier(n_neighbors=3)
+    assert_raises_regex(ValueError,
+                        "Sparse y is only supported for multilabel")
```
You need to call the function here, i.e. 
```python
assert_raises_regex(ValueError,
                    "Sparse y is only supported for multilabel",
                    clf.fit, csr_matrix(X), y)
```

jnothman commented on sklearn/neighbors/tests/test_neighbors.py:
```diff
@@ -503,6 +504,53 @@ def test_kneighbors_classifier_sparse(n_samples=40,
+def test_sparse_multilabel_y():
+    rng = check_random_state(0)
+    n_features = 2
+    n_samples = 100
```
It doesn't need to be this many.

jnothman commented on sklearn/neighbors/tests/test_neighbors.py:
```diff
@@ -503,6 +504,53 @@ def test_kneighbors_classifier_sparse(n_samples=40,
+    n_output = 3
+
+    X = rng.rand(n_samples, n_features)
+    y = rng.randint(0, 5, (n_samples, n_output))
```
It doesn't need to be random. If it is random, we'd rather a fixed random_state: on occasion this could generate a binary array

massich replied:
I thought that this was taken care by `rng = check_random_state(0)`. I moved the line closer for clarity

jnothman replied:
Indeed it was, thanks.

Commit:
Adress jnothman comments

jnothman commented:
LGTM

Commit:
Adress jnotham comments

jnothman commented:
Please add an entry to the change log at `doc/whats_new.rst`.

Commit:
update whats_new, credit to @dalmia

jnothman commented:
You now have test failures

jnothman commented on sklearn/neighbors/classification.py:
```diff
@@ -374,17 +391,51 @@ def predict(self, X):
-            y_pred = y_pred.ravel()
+            y_pred = sparse.hstack(y_pred_sparse_multilabel)
+
+        else:
```
I approved without reading properly. You have this else clause attaching to a for loop. Which is valid syntax, but not doing what you want it to do. I think that for loop is too long and nested to read, and it would be worth taking code out into a helper function.

massich replied:
If I'm not mistaken, this else actually belongs to [`if self.outputs_2d_ == 'sparse':](https://github.com/scikit-learn/scikit-learn/pull/9059/files#diff-364ed6a16c035d2ac7a77e55f1ab7e3bR386)

jnothman commented on sklearn/neighbors/classification.py:
```diff
@@ -374,17 +391,51 @@ def predict(self, X):
                                 dtype=np.int)
 
-            mode = mode.ravel()
+            # y_pred_sparse_multilabel = np.empty()
```
why did you not keep 
```
if self.outputs_2d_ == 'sparse':
    y_pred_sparse_multilabel = []
```
which were here.

jnothman commented on sklearn/neighbors/classification.py:
```diff
@@ -374,17 +391,51 @@ def predict(self, X):
             else:
                 mode = np.array([weighted_mode(pl, w)[0]
                                  for (pl, w)
-                                 in zip(pred_labels[inliers], weights[inliers])],
```
The code here and above had been deleted in #8096: https://github.com/scikit-learn/scikit-learn/pull/8096/files#diff-364ed6a16c035d2ac7a77e55f1ab7e3bL367

Why did you reinsert it?

jnothman reviewed (commented):
Yes, my +1 was definitely premature!

amueller commented:
merge conflicts and test failure?

Commit:
Fix test_classifiers_sparse_multilabel_y

Commit:
Remove test that no longer makes sense

Commit:
Remove unused import

jnothman commented on sklearn/neighbors/classification.py:
```diff
@@ -154,21 +156,38 @@ def predict(self, X):
+                mode = self._mode(_y[neigh_ind, k].toarray(), weights)
+                # `csr_matrix(mode).T` is a CSC matrix with a single column
+                # so that it's fast to hstack it with the other columns
+                y_pred_sparse_multilabel.append(csr_matrix(mode).T)
```
What we should be doing here is csc_matrix(mode.reshape(-1, 1)). Calling .T on a CSR is unnecessary.

massich replied:
Yes, your proposal is faster.
```Python
import scipy.sparse as sp
import numpy as np

f1 = lambda x: sp.csr_matrix(x).T
f2 = lambda x: sp.csc_matrix(x.reshape(-1,1))
data = [np.random.randint(0,2,1000) for _ in range(10000)]

for f in [f1, f2]:
    %timeit sp.hstack([f(d) for d in data])
```
```
1 loop, best of 3: 3.74 s per loop
1 loop, best of 3: 3.1 s per loop
```

But it brakes some test. I'll try to figure it out.

jnothman commented on sklearn/neighbors/classification.py:
```diff
@@ -364,27 +383,51 @@ def predict(self, X):
-            y_pred[outliers, :] = self.outlier_label
+                # `csr_matrix(y_pred_k).T` is a CSC matrix with a single column
+                # so that it's fast to hstack it with the other columns
+                y_pred_sparse_multilabel.append(csr_matrix(y_pred_k).T)
```
Same here

Commit:
Speed improvement by building a sparse matrix differently

Change csr_matrix(x).T for csc_matrix(x.reshape(-1,1))

Commit:
Fix broken test

Commit:
Add rise error test for sparse-multioutput

massich commented:
I added a test to keep the coverage and I found two things wired. 
First a deprecation warning regarding this [line](https://github.com/massich/scikit-learn/blob/8057/sklearn/neighbors/base.py#L786):
```
sklearn.neighbors.tests.test_neighbors.test_sparse_multioutput_error_raise ... /home/sik/Documents/code/scikit-learn/sklearn/neighbors/base.py:786: DeprecationWarning: elementwise == comparison failed; this will raise an error in the future.
  if np.all(np.union1d(y.data, [0, 1]) == [0, 1]):
ok
```
and second, in order to check `y` we have to make it dense which might be incorrect. (see [this](https://github.com/scikit-learn/scikit-learn/pull/9059/files#diff-c569d578c4c466c2b310da6a2b06e1caR253))

@jnothman, @amueller where do we take it from here?

jnothman commented:
isn't the relevant case handled by is_multilabel()?

On 26 Jun 2017 7:02 pm, "Joan Massich" <notifications@github.com> wrote:

> I added a test to keep the coverage and I found two things wired.
> First a deprecation warning regarding this line
> <https://github.com/massich/scikit-learn/blob/8057/sklearn/neighbors/base.py#L786>
> :
>
> sklearn.neighbors.tests.test_neighbors.test_sparse_multioutput_error_raise ... /home/sik/Documents/code/scikit-learn/sklearn/neighbors/base.py:786: DeprecationWarning: elementwise == comparison failed; this will raise an error in the future.
>   if np.all(np.union1d(y.data, [0, 1]) == [0, 1]):
> ok
>
> and second, in order to check y we have to make it dense which might be
> incorrect. (see this
> <https://github.com/scikit-learn/scikit-learn/pull/9059/files#diff-c569d578c4c466c2b310da6a2b06e1caR253>
> )
>
> @jnothman <https://github.com/jnothman>, @amueller
> <https://github.com/amueller> where do we take it from here?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/9059#issuecomment-311003203>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz63JVK3xfkz_GQHgQ7uOr8HhzRaDlks5sH3O7gaJpZM4N0Bnb>
> .
>

massich commented:
I don't think so. But I'm really confused, at this point. 
 `is_multilabel` returns false since `np.unique(y.data)`  is `array([1, 2], dtype=int64)`.

I would expect `type_of_target(y_train)` return something `multilabel-multioutput`. But maybe I'm just constructing `y` plain wrong.

jnothman commented:
We currently don't support multiclass-multioutput with sparse Y. But maybe
that is a silly distinction. And it doesn't make sense to support sparse
multilabel data unless one of the label indicators is 0. So I'd say
something sparse with [1, 2] should be called 'unknown'

On 26 June 2017 at 22:27, Joan Massich <notifications@github.com> wrote:

> I don't think so. But I'm really confused, at this point.
> is_multilabel returns false since np.unique(y.data) is array([1, 2],
> dtype=int64).
>
> I would expect type_of_target(y_train) return something
> multilabel-multioutput. But maybe I'm just constructing y plain wrong.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/9059#issuecomment-311044802>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz69f5E65NtaFx4-rJOTDGm8B3XGsEks5sH6PEgaJpZM4N0Bnb>
> .
>

massich commented:
The only thing I was trying to do was to cover [this line](https://github.com/massich/scikit-learn/blob/8057/sklearn/neighbors/base.py#L789). Because it was not cover. I was not planning to support `multiclass-multioutput`. Just trying to build an example that fall in such category so that the `raise` is triggered.

raghavrv commented on doc/whats_new.rst:
```diff
@@ -212,6 +213,8 @@ Enhancements
      passing a range of bytes to :func:`datasets.load_svmlight_file`.
      :issue:`935` by :user:`Olivier Grisel <ogrisel>`.
+   - Add support for sparse multilabel ``y`` in :class:`NeighborsBase`
+     :issue:`8057` by :user:`Aman Dalmia <dalmia>`.
```
Please do add your name too.

raghavrv commented on sklearn/neighbors/tests/test_neighbors.py:
```diff
@@ -507,6 +508,61 @@ def test_kneighbors_classifier_sparse(n_samples=40,
+    knn_mo = neighbors.KNeighborsClassifier(weights=None,
+                                            algorithm='auto')
+    # print(y_sparse.A)
+    msg = 'Sparse y is only supported for multilabel'
```
Ah okay. so you need to determine if it's multilabel or multioutput based on the shape of `y`. Only then you will be able to raise the appropriate error. Maybe you should modify `type_of_target`?

raghavrv replied:
Humm sorry so multioutput and multilabel are currently of same shape? `(n_samples, n_outputs/n_labels)` which means it's not possible to distinguish between them.

raghavrv replied:
@jnothman thoughts?

jnothman replied:
type_of_target and this code distinguish between them on the basis of number of distinct values.

raghavrv replied:
Thx!

Hmm so you won't be able to distinguish between a binary multioutput and multilabel right? Only if it's non-binary multioutput and multilabel we raise an error?

raghavrv replied:
@massich So you should be making a y which is clearly multioutput (non-binary) so this error can be checked.

jnothman replied:
Binary multioutput is identical to multilabel (semantically).

Commit:
Remove unnecessary comment

Commit:
Add massich as PR contributor

Commit:
Merge branch 'master' into 8057

jnothman commented on doc/whats_new.rst:
```diff
@@ -138,6 +138,18 @@ Model selection and evaluation
      :class:`model_selection.RepeatedStratifiedKFold`.
      :issue:`8120` by `Neeraj Gangwar`_.
 
+   - :func:`metrics.matthews_corrcoef` now support multiclass classification.
```
Merge not quite right.

Commit:
update whats new

Commit:
Merge remote-tracking branch 'sklearn/master'

Commit:
xx

Commit:
add sparse multilabel support to whats new

Commit:
catch the case of sparse multilabel multioutput

massich commented on sklearn/utils/multiclass.py:
```diff
@@ -247,7 +247,10 @@ def type_of_target(y):
-        y = np.asarray(y)
+        if not issparse(y):
+            y = np.asarray(y)
+        else:
```
When it is sparse multilabel multioutput

massich replied:
which is the case to raise an error.

jnothman replied:
I think you must mean sparse multiclass multioutput.

jnothman replied:
Well, I'd still like this to return unknown seeing as we don't support sparse arrays as a representation for multiclass multioutput... I was just trying to clarify your use of terminologies.

massich commented:
sorry for the cryptic commit messages.

jnothman commented on sklearn/utils/tests/test_multiclass.py:
```diff
@@ -140,6 +140,9 @@ def __array__(self, dtype=None):
         # 3d
         np.array([[[0, 1], [2, 3]], [[4, 5], [6, 7]]]),
+
+        # sparse multilabel-multioutput
```
multiclass-multioutput.

Multilabel == binary multioutput.

jnothman commented on sklearn/neighbors/tests/test_neighbors.py:
```diff
@@ -507,6 +508,61 @@ def test_kneighbors_classifier_sparse(n_samples=40,
+                                                        random_state=0)
+    clf.fit(X_train, y_train)
+    y_sparse_pred = clf.predict(X_test)
+    assert_array_equal(y_pred, y_sparse_pred.toarray())
```
please assert about `y_sparse_pred.format`

jnothman reviewed (commented):
we should update the type spec in the docstring for fit's y and predict's return value.

jnothman commented:
Otherwise LGTM

Commit:
fix minor things based on Joel's feedback

massich commented on sklearn/neighbors/base.py:
```diff
@@ -776,16 +776,33 @@ def fit(self, X, y):
-        check_classification_targets(y)
+        try:
+            check_classification_targets(y)
+        except ValueError as e:
```
I find this part of the code ugĺy. Even more now that we added `'sparse-muliclass-multioutput'` as a `type_of_target(y)` [here](https://github.com/scikit-learn/scikit-learn/pull/9059/files#diff-c569d578c4c466c2b310da6a2b06e1caR252). Wouldn't it be a way to make use of the message from type_of_target to avoid checking again `issparse(y) and self.outputs_2d_`  ? 
cc: @jnothman, @ogrisel

jnothman reviewed (commented):
I agree that the ugliness here is a major deterrent. Anything you think tidies it up is welcome, IMO.

massich commented:
One would be custom errors, but I think this is not a path we want to take. The other option is to add something to `check_classification_targets` signature. i.e:

```py
def check_classification_targets(y, complementary_msg=''):
     ....
             raise ValueError("Unknown label type: %r.\n%r" % (y_type, complementary_msg))
```

So that the call would be: 
```diff
- try:
-     check_classification_targets(y)
- except ValueError as e:
-     if issparse(y) and self.outputs_2d_:
-         raise ValueError("Sparse y is only supported for multilabel"
-                           "case (multioutput is not supported)")
-     else:
-         raise e
+     check_classification_targets(y, complementary_msg=("Sparse y is only supported"
+          "for multilabel case (multioutput is not supported)")
```

Commit:
Update tests to use 'sparse-multiclass-multioutput'

Commit:
Merge branch 'master' into 8057

Commit:
ensure y_pred CSC type

jnothman commented on sklearn/neighbors/base.py:
```diff
@@ -759,6 +759,11 @@ def fit(self, X, y):
             Target values of shape = [n_samples] or [n_samples, n_outputs]
+            If y is sparse with values [0, 1], it is expected to be the
+            indicator representation of a multilabel classification problem.
+            Otherwise, if y is an array with integer dtype, it is interpreted
```
y can have non-integer dtype in the multiclass case.

Better to just say that "A sparse matrix may only be used for multilabel targets."

jnothman reviewed (commented):
Otherwise this still LGTM!

Commit:
Merge branch 'master' into 8057

Commit:
update whats_new + minor changes

Commit:
Merge branch 'master' into 8057

glemaitre commented on sklearn/neighbors/classification.py:
```diff
@@ -154,21 +156,38 @@ def predict(self, X):
-        if not self.outputs_2d_:
-            y_pred = y_pred.ravel()
+            # Old versions of scipy hstack returns COO formatted matrix
+            y_pred = sparse.hstack(y_pred_sparse_multilabel).tocsc()
```
note that this is for readability, the performance will be the same

glemaitre commented on sklearn/neighbors/classification.py:
```diff
@@ -364,27 +383,51 @@ def predict(self, X):
 
-            mode = mode.ravel()
+            for k, classes_k in enumerate(classes_):
+                pred_labels = np.array([_y[ind, k].toarray()
```
why this is converted to `object`. if we have a sparse matrix, we have for sure numerical values.
Then we don't even have to construct a dense matrix. we can just pass `*.data` from the sparse matrix

glemaitre replied:
In this case we need also to slice the weights

jnothman replied:
So this is extracting a single column for all of the neighbors of each query. We are making it dense here, at a cost of (n_samples * avg_neighbors) memory. The sparse representation is to avoid a dense matrix of (n_samples * n_classes) where n_classes is presumed to be large. Yes, we can avoid making this dense, but I don't think it's especially problematic as long as avg_neighbours << n_classes.

We should be careful about constructing arrays of arrays/lists though. When the lengths of the arrays are all the same, this becomes a 2d array whose elements are ints, but still of dtype object due to this constructor. The safe way to do such construction is with np.empty(n, dtype='O') and then fill the array. But I don't see why pred_labels should be an array here at all.

Now that I've clarified that to myself, I'll go see what https://github.com/massich/scikit-learn/pull/5.

massich replied:
Conclusion? we materialize the matrix here and later we can modify the mode computation propagating the sparse matrix later if needed. 
If so, we could use `mode = 1 if ( np.product(x.shape) - x.nnz ) < x.nnz else 0`

jorisvandenbossche replied:
> We should be careful about constructing arrays of arrays/lists though.

That is correct, so in the unlikely cornercase that each sample has the same number of neighbours, the current code is probably broken.

> But I don't see why pred_labels should be an array here at all.

The reason is because further down below, we use fancy numpy indexing with a list on this array, which does not work for a list (in `pred_labels[inliers]`). That can of course be converted to list comprehension, but might have performance bottleneck (not sure this is the case for object array)

jorisvandenbossche replied:
Actually, the ` zip(pred_labels[inliers], weights[inliers])]` works fine as well if it is a 2D array instead of 1D object array? So I think we can leave this as is.

glemaitre commented on sklearn/neighbors/classification.py:
```diff
@@ -364,27 +383,51 @@ def predict(self, X):
+        else:
+            y_pred = np.empty((n_samples, n_outputs), dtype=classes_[0].dtype)
+            for k, classes_k in enumerate(classes_):
+                pred_labels = np.array([_y[ind, k] for ind in neigh_ind],
```
I have the same concern than ealier regarding the sparse matrix

massich replied:
this one is not sparse.

glemaitre commented on sklearn/neighbors/classification.py:
```diff
@@ -364,27 +383,51 @@ def predict(self, X):
+
+    def _mode(self, pred_labels, weights, inliers):
+        if weights is None:
+            mode = np.array([stats.mode(pl)[0]
```
why using a list comprehension and not using `axis` option from `stats.mode` or `weighted_mode`?

glemaitre commented on sklearn/neighbors/classification.py:
```diff
@@ -364,27 +383,51 @@ def predict(self, X):
 
         return y_pred
+
+    def _mode(self, pred_labels, weights, inliers):
```
move the function above it is used.

glemaitre replied:
I would not pass inliers because we take inliers from weights but not from pred_labels. It seems inconsistant.

we should have either `pred_labels[inliers], weight[inliers]` or `pred_inliers, weight_inliers` passed as argument.
I am more in favor of the second because it will be easier with sparse matrices.

glemaitre reviewed (changes requested):
without checking the tests for the moment

massich commented on doc/whats_new.rst:
```diff
@@ -36,6 +36,8 @@ Classifiers and regressors
 
+- Add support for sparse multilabel ``y`` in :class:`NeighborsBase`
+  :issue:`8057` by :user:`Aman Dalmia <dalmia>`, :user:`Joan Massich <massich>`.
 - Added :class:`naive_bayes.ComplementNB`, which implements the Complement
```
it is under 0.20

Commit:
Improve hstack call to make it more radable

Commit:
Add is_valid_classification_target(y) to improve error treatment

The initial idea with @ogrisel was to change the ValueError's meassage for
something more descriptive like this:

```py
  raise ValueError("Unknown classification label type. Got: %r" % y)
```

The main problem is that such message breaks this:

```sh
  $pytest sklearn/tests/test_common.py::test_non_meta_estimators
```

Commit:
Move _mode function before its called

Commit:
Merge branch 'master' into 8057_sklearn_pr

massich commented:
When addressing [this](https://github.com/scikit-learn/scikit-learn/pull/9059/#discussion_r136551392) comment by @glemaitre, I stumble into some problems. I don't really know how to post a concrete question, so I collected my thoughts [here as a PR to this PR](
https://github.com/massich/scikit-learn/pull/5) and I would love to get some feedback. Thx.

cc: @jnothman, @lesteve

massich commented on sklearn/neighbors/base.py:
```diff
@@ -776,15 +776,14 @@ def fit(self, X, y):
+                                 " supported). Got: %r" % y)
             else:
-                raise
+                raise ValueError("Unknown label type: %r" % y)
```
The initial idea with @ogrisel was to change the ValueError's meassage for something more descriptive like this:

```py
  raise ValueError("Unknown classification label type. Got: %r" % y)
```

The main problem is that such message breaks this:

```sh
  $pytest sklearn/tests/test_common.py::test_non_meta_estimators
```
If needed I'll change the message in a separated PR, so that no common files are changed here.

lesteve commented on doc/whats_new.rst:
```diff
@@ -4,6 +4,5758 @@
 Release History
 ===============
+
+Version 0.20 (under development)
```
You want to add your entry to v0.20.0.rst (or something like this) rather than change whats_new.rst.

glemaitre commented:
@massich Could you solve the conflict and address the remaining comments

massich committed:
Merge branch 'master' into 8057

massich commented:
@dorcoh feel free to take over this one! let me know if you need anything

ogrisel commented on sklearn/neighbors/classification.py:
```diff
@@ -154,21 +156,38 @@ def predict(self, X):
+            mode, _ = weighted_mode(neigh, weights, axis=1)
+
+        mode = np.asarray(mode.ravel(), dtype=np.intp)
+        return mode
```
Style:

```python
return np.asarray(mode.ravel(), dtype=np.intp)
```

Commit:
Merge branch 'master' into 8057
